ddsrouter_utils/include/ddsrouter_utils/memory/OwnerPtr.hpp

```diff
@@ -0,0 +1,320 @@
+namespace ddsrouter {
+namespace utils {
+
+template <class T>
+class OwnerPtr;
+
+/**
+ * @brief This classes implement a new smart pointer that allows to reference data with only one ownership
+ * from different scopes and threads.
+ *
+ * INSTRUCTIONS
+ * The correct way of using it is having a \c OwnerPtr object referencing the data, and getting
+ * a lessee from it.
+ * Before using the data in the ptr, use \c lock and check the result is not null.
+ * Destroy the ptr after using it. It is advisable to destroy it soon, as the owner could want to erase it.
+ *
+ * USE CASE
+ * Whenever an object must be created and deleted from a single point, but could be used from different
+ * scopes and/or threads.
+ *
+ * WHY NEW CLASS
+ * It is similar to a shared_ptr and weak_ptr, but this assures the object is only destroyed from its
+ * owner, and only if any lessee is not using it at the moment.
+ *
+ * IMPLEMENTATION
+ * The way it works, is that every \c OwnerPtr shares a mutex with every \c LesseePtr , so while the
+ * data is being used by a lessee, the owner could not destroy it.
+ * Whenever the object is used from a lessee, a new ptr is created and mutex lock. When this ptr is deleted
+ * the mutex is unlocked (by unique_ptr deleter).
+ *
+ *
+ * EXAMPLE OF USAGE
+ * OwnerPtr<T> owner(new T());
+ * LesseePtr<T> lessee = owner.lease();
+ * {
+ *   auto locked_ptr_ = lessee.lock(); // From here till locked_ptr_ destruction, data could not be destroyed.
+ *   if (locked_ptr_) {
+ *     locked_ptr_->foo(); // Use data without worrying it could be destroyed
+ *   } else {
+ *     // Ups, the data is not available anymore
+ *   }
+ * }
+ */
+
+/**
+ * This class contains a reference (ptr) to a value of type T but do not own it.
+ * Not owning the object implies that, when using, this reference could have been invalidated.
+ * For that, method \c lock create a smart pointer referencing that object, that will not be destroyed
+ * while the smart pointer exists.
+ *
+ * @tparam T Type of the data referenced by this ptr.
+ *
+ * @note It could only be created from a \c OwnerPtr object, the owner of the data referenced.
+ */
+template <class T>
+class LesseePtr
+{
+public:
+
+    /**
+     * @brief Construct a new empty Lessee Ptr object
+     *
+     * This constructor is only used to create an empty object.
+     * It could not contain any data.
+     * To use an empty ptr, use operator= .
+     */
+    LesseePtr();
+
+    /**
+     * @brief Destroy the Lessee Ptr object
+     *
+     * It waits in case there are locks taken.
+     * It does not destroy the data (it has not the ownership).
+     */
+    ~LesseePtr();
+
+    /**
+     * @brief copy assigment
+     *
+     * It gets the reference from the other object.
+     * It looses an old reference and the mutex in case it had it.
+     *
+     * @param other object to copy
+     * @return LesseePtr<T>& this object
+     */
+    LesseePtr<T>& operator =(
+            const LesseePtr<T>& other);
```

Comment:
I believe the copy assignment should be replaced by a move one, as the owner pointer should be the only one in charge of creating lessee pointers, with a single mutex being only shared between a single lessee and its owner. Correct me if I'm wrong.